app/src/main/java/com/appdev/eateryblueandroid/util/LoginRepository.kt

```diff
@@ -0,0 +1,295 @@
+                                date.year,
+                                date.month,
+                                date.day,
+                                date.hour,
+                                date.minute
+                            ),
+                            ordinalToTransactionType(transProto.transactionType),
+                            ordinalToAccountType(transProto.accountType),
+                            transProto.location
+                        )
+                    )
+                }
+                transactionsFlow.value = mutableTransactions
+
+                Log.i(
+                    "Caching",
+                    "Transactions History Caching done with " + transactionsFlow.value.size + " items pulled."
+                )
+
+                hasLoadedTransactions = true
+                cachedFlow.value = hasLoadedAccounts
+                attemptAutoLogin(profileViewModel)
+
+                if (transactionsFlow.value.isNotEmpty()) {
+                    this.cancel()
+                }
+            }
+        }
+    }
+
+    /**
+     * Saves user profile data to proto datastore.
+     */
+    fun cacheAccountInfo(accounts: List<Account>, transactions: List<Transaction>) {
+        val accMutable: MutableList<AccountProto> = mutableListOf()
+        accounts.forEach { acc ->
+            accMutable.add(
+                AccountProto.newBuilder().setType(acc.type!!.ordinal).setBalance(acc.balance ?: 0.0)
+                    .build()
+            )
+        }
+        val transactionsMutable: MutableList<TransactionProto> = mutableListOf()
+        transactions.forEach { trans ->
+            // Some transactions (like laundry depositing for some reason??) have no date.
+            // Not important though since those transactions are not displayed anyways,
+            // so just put some default date:
+            val date: LocalDateTime = trans.date ?: LocalDateTime.of(1, 1, 1, 1, 1)
+
+            transactionsMutable.add(
+                TransactionProto.newBuilder()
+                    .setId(trans.id)
+                    .setAmount(trans.amount!!)
+                    .setResultingBalance(trans.resultingBalance ?: 0.0)
+                    .setDate(
+                        Date.newBuilder()
+                            .setYear(date.year)
+                            .setMonth(date.month.value)
+                            .setDay(date.dayOfMonth)
+                            .setHour(date.hour)
+                            .setMinute(date.minute)
+                            .build()
+                    )
+                    .setTransactionType(trans.transactionType!!.ordinal)
+                    .setAccountType(trans.accountType!!.ordinal)
+                    .setLocation(trans.location!!)
+                    .build()
+            )
+        }
+
+        CoroutineScope(Dispatchers.IO).launch {
+            appContext!!.userPreferencesStore.updateData { currentPreferences ->
```

Comment:
These updates are more suitable for a [UserPreferencesRepository](https://github.com/cuappdev/volume-compose-android/blob/chris/Hilt/app/src/main/java/com/cornellappdev/volume/data/repositories/UserPreferencesRepository.kt) that deals strictly with interacting with the data store source

Reply:
I was considering doing this at some point, but it feels more comfortable to leave save and load operations in the respective repositories that they pertain to. For example, in LoginRepository.kt I use saveLoginInfo to change the repository data before saving to local storage. If I had the save operation in a UserPreferencesRepository, I believe it would be harder to coordinate this dual function.

However, I may be misinterpreting you--do you mean putting **only** the update data / coroutine part into a UserPreferencesRepository? And then having the repositories communicating between each other? If so, I wouldn't be opposed to implementing this.

Reply:
I think it is fine for now actually. There are some architectural issues that go deeper that I will try and address in another PR in hopes of making the codebase easier to navigate!